File: attach-the-payment-method/node/attach-payment-method.js
/**
 * Attach a WooCommerce customer's saved Stripe PaymentMethod to their Stripe
 * Customer when it exists but was never attached. Run on a schedule, ahead
 * of billing. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/attach-the-payment-method/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 30);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

/** The saved Stripe PaymentIntent id lives on meta _stripe_intent_id. We use
 * it to look up the PaymentIntent, then read payment_method off it. Some
 * older orders only have a pm_ id directly on transaction_id.
 */
export function paymentMethodIdOf(order) {
  const tid = order.transaction_id || "";
  if (tid.startsWith("pm_")) return tid;
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  return null;
}

async function stripeCustomerIdOf(wcCustomerId) {
  if (!wcCustomerId) return null;
  const customer = await woo(`/customers/${wcCustomerId}`);
  for (const meta of customer?.meta_data || []) {
    if (meta.key === "_stripe_customer_id") return meta.value;
  }
  return null;
}

/** Turn whatever id we found (a pi_... or a pm_...) into a PaymentMethod
 * object read straight from Stripe. Returns null if it cannot be resolved.
 */
async function resolvePaymentMethod(rawId) {
  if (!rawId) return null;
  try {
    if (rawId.startsWith("pi_")) {
      const intent = await stripe.paymentIntents.retrieve(rawId);
      return intent.payment_method ? stripe.paymentMethods.retrieve(intent.payment_method) : null;
    }
    return await stripe.paymentMethods.retrieve(rawId);
  } catch {
    return null;
  }
}

/** Pure decision function. No I/O. Takes the Stripe Customer id the
 * WooCommerce customer should be linked to, and the PaymentMethod object
 * (with a "customer" field) as read from Stripe, and returns [action, reason].
 *
 * Actions:
 *   skip     - nothing to check or nothing to compare against
 *   ok       - already attached to the expected customer, no change needed
 *   conflict - attached to a different customer, needs a human to review
 *   attach   - unattached, safe to attach automatically
 */
export function decide(stripeCustomerId, paymentMethod) {
  if (!paymentMethod) return ["skip", "no PaymentMethod found to check"];
  if (!stripeCustomerId) return ["skip", "customer has no Stripe Customer id on file"];
  const current = paymentMethod.customer || null;
  if (current === stripeCustomerId) return ["ok", "already attached to the right customer"];
  if (current) return ["conflict", `attached to a different customer (${current})`];
  return ["attach", "unattached, safe to attach"];
}

async function attachPaymentMethod(paymentMethodId, stripeCustomerId) {
  await stripe.paymentMethods.attach(paymentMethodId, { customer: stripeCustomerId });
}

async function addNote(orderId, note) {
  await woo(`/orders/${orderId}/notes`, { method: "POST", body: JSON.stringify({ note }) });
}

async function* recentOrders() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/orders?after=${after}&per_page=50&page=${page}`);
    if (!batch || !batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

export async function run() {
  let fixed = 0;
  for await (const order of recentOrders()) {
    const rawId = paymentMethodIdOf(order);
    if (!rawId) continue;
    const paymentMethod = await resolvePaymentMethod(rawId);
    const stripeCustomerId = await stripeCustomerIdOf(order.customer_id);
    const [action, reason] = decide(stripeCustomerId, paymentMethod);
    if (action === "conflict") {
      console.warn(`Order ${order.id}: ${reason}. Needs a human to review.`);
      continue;
    }
    if (action === "skip" || action === "ok") continue;
    const pmId = paymentMethod.id;
    console.log(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would attach" : "attaching"}`);
    if (!DRY_RUN) {
      await attachPaymentMethod(pmId, stripeCustomerId);
      await addNote(
        order.id,
        `Attached Stripe PaymentMethod ${pmId} to Stripe Customer ${stripeCustomerId}. ` +
          `It existed but was not attached, which would have blocked the next off session charge.`
      );
    }
    fixed++;
  }
  console.log(`Done. ${fixed} PaymentMethod(s) ${DRY_RUN ? "to attach" : "attached"}.`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: attach-the-payment-method/python/attach_payment_method.py
"""Attach a WooCommerce customer's saved Stripe PaymentMethod to their Stripe
Customer when it exists but was never attached. Run on a schedule, ahead of
billing. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/attach-the-payment-method/
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("attach_payment_method")

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "30"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def payment_method_id_of(order):
    """The saved Stripe PaymentIntent id lives on meta _stripe_intent_id.
    We use it to look up the PaymentIntent, then read payment_method off it.
    Some older orders only have a pm_ id directly on transaction_id.
    """
    tid = order.get("transaction_id") or ""
    if tid.startswith("pm_"):
        return tid
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    return None


def stripe_customer_id_of(wc_customer_id):
    if not wc_customer_id:
        return None
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/customers/{wc_customer_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    for meta in r.json().get("meta_data") or []:
        if meta.get("key") == "_stripe_customer_id":
            return meta.get("value")
    return None


def resolve_payment_method(raw_id):
    """Turn whatever id we found (a pi_... or a pm_...) into a PaymentMethod
    object read straight from Stripe. Returns None if it cannot be resolved.
    """
    if raw_id is None:
        return None
    try:
        if raw_id.startswith("pi_"):
            intent = stripe.PaymentIntent.retrieve(raw_id)
            pm_id = intent.get("payment_method")
            return stripe.PaymentMethod.retrieve(pm_id) if pm_id else None
        return stripe.PaymentMethod.retrieve(raw_id)
    except stripe.error.InvalidRequestError:
        return None


def decide(stripe_customer_id, payment_method):
    """Pure decision function. No I/O. Takes the Stripe Customer id the
    WooCommerce customer should be linked to, and the PaymentMethod object
    (a plain dict-like with a "customer" field) as read from Stripe, and
    returns (action, reason).

    Actions:
      skip     - nothing to check or nothing to compare against
      ok       - already attached to the expected customer, no change needed
      conflict - attached to a different customer, needs a human to review
      attach   - unattached, safe to attach automatically
    """
    if payment_method is None:
        return ("skip", "no PaymentMethod found to check")
    if not stripe_customer_id:
        return ("skip", "customer has no Stripe Customer id on file")
    current = payment_method.get("customer")
    if current == stripe_customer_id:
        return ("ok", "already attached to the right customer")
    if current:
        return ("conflict", f"attached to a different customer ({current})")
    return ("attach", "unattached, safe to attach")


def attach_payment_method(payment_method_id, stripe_customer_id):
    stripe.PaymentMethod.attach(payment_method_id, customer=stripe_customer_id)


def add_note(order_id, note):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": note}, auth=AUTH, timeout=30,
    ).raise_for_status()


def recent_orders():
    page = 1
    after = f"{__import__('datetime').date.today() - __import__('datetime').timedelta(days=LOOKBACK_DAYS)}T00:00:00"
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"after": after, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def run():
    fixed = 0
    for order in recent_orders():
        raw_id = payment_method_id_of(order)
        if raw_id is None:
            continue
        payment_method = resolve_payment_method(raw_id)
        stripe_customer_id = stripe_customer_id_of(order.get("customer_id"))
        action, reason = decide(stripe_customer_id, payment_method)
        if action == "conflict":
            log.warning("Order %s: %s. Needs a human to review.", order["id"], reason)
            continue
        if action in ("skip", "ok"):
            continue
        pm_id = payment_method["id"]
        log.info("Order %s: %s. %s", order["id"], reason, "would attach" if DRY_RUN else "attaching")
        if not DRY_RUN:
            attach_payment_method(pm_id, stripe_customer_id)
            add_note(
                order["id"],
                f"Attached Stripe PaymentMethod {pm_id} to Stripe Customer "
                f"{stripe_customer_id}. It existed but was not attached, which "
                f"would have blocked the next off session charge.",
            )
        fixed += 1
    log.info("Done. %d PaymentMethod(s) %s.", fixed, "to attach" if DRY_RUN else "attached")


if __name__ == "__main__":
    run()
